### PR TITLE
Improve Get-VSwitchPathCost.ps1 performance

### PR DESCRIPTION
**Pull request by tmolenh**
Instead of collecting each counter in its own Job, collect them all at once because it much faster.

### CODE REVIEW

**Comment by omarcardona** [comment]
I'll leave the Squash and Merge steps to Khoa.  Please make sure he reviews before final integration.

File: Diagnostics/Get-VSwitchPathCost.ps1 [hunk 1]
```diff
@@ -127,34 +136,25 @@ function Get-CounterAverage([String[]] $statArray)
 
 if (IsHyperThreadingEnabled)
 {
-    Log "ERROR: Hyper-Threading is enabled. Please disable Hyper-Threading before starting the test."
+    Log "WARNING: Hyper-Threading is enabled. This may affect the accuracy of the results."
 }
 
-$switch = Get-VMSwitch $SwitchName
```
Comment (tmolenh) [inline Diagnostics/Get-VSwitchPathCost.ps1]: This cmdlet is the only part of the script that requires admin, so I removed it in favor of checking if the counters exist. However, I don't check for duplicate switch names now.

File: Diagnostics/Get-VSwitchPathCost.ps1 [hunk 2]
```diff
@@ -127,34 +136,25 @@ function Get-CounterAverage([String[]] $statArray)
...
-    $MaxCpuNumber = $cpuPerNuma * $numaCount - 1
+    $MaxCpuNumber = Get-MaxCpuNumber
     Log "Set MaxCpuNumber to $MaxCpuNumber"
 }
-
-if ($MaxCpuNumber -lt $BaseCpuNumber)
+elseif (IsHyperThreadingEnabled)
 {
-   Log "ERROR: Max CPU number is less than base CPU number"
+    # force MaxCpuNumber to odd value with HT so the entire core is measured
```
Comment (tmolenh) [inline Diagnostics/Get-VSwitchPathCost.ps1]: This is correct, right?

File: Diagnostics/Get-VSwitchPathCost.ps1 [hunk 3]
```diff
@@ -1,74 +1,52 @@
...
    This script does not start any traffic. It is the caller's responsibility to ensure
    traffic is already running in steady state, before calling the script.
    The script only collects perfmon counters for the specified during, and computes
    path cost, based on the specified CPU range.
 
-   Since CPU utilization and path cost are more accurately measured with HyperThreading
-   disabled, this script forces HyperThreading to be disabled. You can disable HyperThreading
-   in the BIOS.
-
    However, note that for better performance, HyperThreading should be enabled.
```
Comment (mskhot) [inline Diagnostics/Get-VSwitchPathCost.ps1]: should remove this line, if you remove the paragraph above?

File: Diagnostics/Get-VSwitchPathCost.ps1 [hunk 4]
```diff
@@ -88,98 +66,82 @@ function Log($Message)
     }
 }
 
+$cpuInfo = @(Get-WmiObject -Class win32_processor -Property NumberOfCores, MaxClockSpeed, NumberOfLogicalProcessors)
```
Comment (mskhot) [inline Diagnostics/Get-VSwitchPathCost.ps1]: Can you check to see if there's a property for currentClockSpeed, and throw an error if MaxClockSpeed != CurrentClockSpeed?  In the BIOS, by default, the machine uses a PowerSaving profile, and the CPU speed is much lower than current clock speed.

We want the power profile to be "Performance", when doing perf measurements, so the CPU speed doesn't vary.

File: Diagnostics/Get-VSwitchPathCost.ps1 (hunk 2 above)
Comment (mskhot) [reply Diagnostics/Get-VSwitchPathCost.ps1]: No, with HyperThreading enabled, we can't really deterministically measure the CPU cycles. 
Also, the maxCpuNumber calculation below is not correct.  This means you're ignoring half of the CPUs.  The HT core pairs are 0/1, 2/3, 4/5, etc.

File: Diagnostics/Get-VSwitchPathCost.ps1 [hunk 5]
```diff
@@ -1,293 +1,254 @@
...
-}
-
-#
-# Verify HyperThreading is disabled
-#
-if (IsHyperThreadingEnabled)
+elseif (IsHyperThreadingEnabled)
 {
-    Log "ERROR: HyperThreading is enabled. Please disable HyperThreading before starting the test."
+    # force MaxCpuNumber to odd value with HT so the entire core is measured
```
Comment (mskhot) [inline Diagnostics/Get-VSwitchPathCost.ps1]: maybe print a warning?

Comment (tmolenh) [reply Diagnostics/Get-VSwitchPathCost.ps1]: Just realized that BaseCpuNumber should be forced to even value as well.

**Pull request merged**